File: sync-core/src/main/java/dev/thompgt/habitsync/sync/WireChange.java
package dev.thompgt.habitsync.sync;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

/**
 * A {@link Change} in its over-the-wire form: enums as strings, the HLC in its compact
 * encoding, field values as plain text.
 *
 * <p>Deliberately not the same type as {@code Change}. The wire format is a compatibility
 * surface that outlives any single client version, and keeping it separate means an
 * unknown enum value from a newer client is a validation error the server can answer with
 * a 400, rather than a deserialisation crash. It also means {@code Change} stays free to
 * gain type-safe structure without breaking every deployed device.
 *
 * <p>This record lives in sync-core, not in the server, so that both ends of the protocol
 * encode against one definition. When the server owned its own copy, the two could drift
 * in exactly the way that is hardest to notice — a field one side writes and the other
 * silently ignores, discovered as missing data weeks later.
 *
 * <h2>Null field values are load-bearing</h2>
 *
 * {@code fields} permits null <em>values</em>, and the map is therefore not built with
 * {@link Map#copyOf}, which rejects them. A present key with a null value means "clear
 * this field"; an absent key means "this change does not touch it". Conflating the two
 * turns every clear operation into a silent no-op — see {@link FieldValue}.
 *
 * @param opId       client-generated, stable across retries; the basis of push idempotency
 * @param entityType {@link EntityType} name
 * @param entityId   client-generated entity id
 * @param kind       {@link OpKind} name
 * @param hlc        {@link Hlc#toCompactString()} form
 * @param fields     field writes for {@code UPSERT}; empty otherwise
 */
public record WireChange(
        UUID opId,
        String entityType,
        UUID entityId,
        String kind,
        String hlc,
        Map<String, String> fields) {

    public WireChange {
        // No null checks and no Map.copyOf here: this type is populated by a JSON parser
        // from untrusted input, and a half-built instance must be constructible so that
        // ChangeCodec can report *which* part is wrong. Validation belongs in decode.
        fields = fields == null
                ? Map.of()
                : Collections.unmodifiableMap(new LinkedHashMap<>(fields));
    }

    /** The change this describes, for logs and error messages. */
    @Override
    public String toString() {
        return "%s %s/%s @%s".formatted(kind, entityType, entityId, hlc);
    }

}

File: ios/Sources/HabitSyncCore/WireChange.swift
import Foundation

/// A ``Change`` in its over-the-wire form: enums as strings, the HLC in its compact
/// encoding, field values as plain text.
///
/// Deliberately not the same type as ``Change``. The wire format is a compatibility surface
/// that outlives any single client version, and keeping it separate means an unknown enum
/// value from a newer peer is a validation error this client can report, rather than a
/// decoding crash. It also means ``Change`` stays free to gain type-safe structure without
/// breaking every deployed device.
///
/// ## Null field values are load-bearing
///
/// `fields` permits null *values*. A present key with a null value means "clear this
/// field"; an absent key means "this change does not touch it". Conflating the two turns
/// every clear operation into a silent no-op — see ``FieldValue``.
///
/// The `Codable` conformance below is written out rather than synthesised for exactly that
/// reason. Swift's default handling of optional values inside a dictionary is easy to get
/// subtly wrong (`dict[key] = nil` *removes* the key rather than storing a null), and the
/// failure is invisible: the request encodes, the server accepts it, and the user's cleared
/// field just never travels.
public struct WireChange: Hashable, Sendable, Codable {

    /// Client-generated, stable across retries; the basis of push idempotency.
    public let opId: UUID
    /// ``EntityType`` raw value.
    public let entityType: String
    /// Client-generated entity id.
    public let entityId: UUID
    /// ``OpKind`` raw value.
    public let kind: String
    /// ``Hlc/compactString`` form.
    public let hlc: String
    /// Field writes for `UPSERT`; empty otherwise. A present key with a nil value is a clear.
    public let fields: [String: String?]

    public init(
        opId: UUID,
        entityType: String,
        entityId: UUID,
        kind: String,
        hlc: String,
        fields: [String: String?]
    ) {
        self.opId = opId
        self.entityType = entityType
        self.entityId = entityId
        self.kind = kind
        self.hlc = hlc
        self.fields = fields
    }

    private enum Key: String, CodingKey {
        case opId, entityType, entityId, kind, hlc, fields
    }

    /// A coding key whose name is only known at runtime — one per field written.
    private struct FieldKey: CodingKey {
        let stringValue: String
        var intValue: Int? { nil }
        init(_ name: String) { stringValue = name }
        init?(stringValue: String) { self.stringValue = stringValue }
        init?(intValue: Int) { return nil }
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: Key.self)
        opId = try container.decode(UUID.self, forKey: .opId)
        entityType = try container.decode(String.self, forKey: .entityType)
        entityId = try container.decode(UUID.self, forKey: .entityId)
        kind = try container.decode(String.self, forKey: .kind)
        hlc = try container.decode(String.self, forKey: .hlc)

        // An absent map and an empty one mean the same thing, and a DELETE legitimately
        // carries neither.
        if container.contains(.fields), try !container.decodeNil(forKey: .fields) {
            let raw = try container.nestedContainer(keyedBy: FieldKey.self, forKey: .fields)
            var decoded: [String: String?] = [:]
            for key in raw.allKeys {
                if try raw.decodeNil(forKey: key) {
                    // updateValue, not subscript assignment: `decoded[k] = nil` deletes the
                    // key, which is exactly the clear-versus-untouched conflation this type
                    // exists to prevent.
                    decoded.updateValue(nil, forKey: key.stringValue)
                } else {
                    decoded.updateValue(try raw.decode(String.self, forKey: key), forKey: key.stringValue)
                }
            }
            fields = decoded
        } else {
            fields = [:]
        }
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: Key.self)
        try container.encode(opId, forKey: .opId)
        try container.encode(entityType, forKey: .entityType)
        try container.encode(entityId, forKey: .entityId)
        try container.encode(kind, forKey: .kind)
        try container.encode(hlc, forKey: .hlc)

        var raw = container.nestedContainer(keyedBy: FieldKey.self, forKey: .fields)
        // Sorted purely so a request body is reproducible in a test or a packet capture;
        // JSON object order carries no meaning to either end.
        for name in fields.keys.sorted() {
            guard let value = fields[name] else { continue }
            if let value {
                try raw.encode(value, forKey: FieldKey(name))
            } else {
                try raw.encodeNil(forKey: FieldKey(name))
            }
        }
    }
}
